### PR TITLE
Add Distinct.

### PR DESCRIPTION
**Pull request by gmpl**

### CODE REVIEW

File: src/Core/QueryT.cs [hunk 1]
```diff
@@ -94,6 +94,9 @@ select GetResult(context).Select(e => e.Value)
 
         public Query<T> Take(int count) =>
             Bind(xs => Query.Return(xs.Take(count)));
+        
+        public Query<T> Distinct =>
```
Comment (atifaziz) [inline src/Core/QueryT.cs]: Turn this into a method and add an overload that enables one to supply an equality comparer, just as you have with [`Enumerable.Distinct`](https://msdn.microsoft.com/en-us/library/bb338049(v=vs.110).aspx).

File: src/Core/GenericEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Comment (atifaziz) [inline src/Core/GenericEqualityComparer.cs]: - License header is missing
- Imports should go inside the namespace declaration below

File: src/Core/GenericEqualityComparer.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+
+namespace WebLinq
+{
+    public class GenericEqualityComparer<T, U> : IEqualityComparer<U>
+    {
+        private Func<U, T> contraMapper;
```
Comment (atifaziz) [inline src/Core/GenericEqualityComparer.cs]: `private` is redundant

File: src/Core/GenericEqualityComparer.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace WebLinq
+{
+    public class GenericEqualityComparer<T, U> : IEqualityComparer<U>
```
Comment (atifaziz) [inline src/Core/GenericEqualityComparer.cs]: - No need to call it generic. It's obvious so rename class to simply `EqualityComparer`
- Missing `sealed`
- `U` is unconventional

File: src/Core/GenericEqualityComparer.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        private Func<U, T> contraMapper;
+        private IEqualityComparer<T> comparer;
+
+        public GenericEqualityComparer(Func<U, T> contraMapper, IEqualityComparer<T> comparer)
```
Comment (atifaziz) [inline src/Core/GenericEqualityComparer.cs]: If you reverse the argument, you could turn this entirely into an extension method and move the class as a private implementation detail of the _extending class_.

File: src/Core/Comparer.cs [hunk 6]
```diff
@@ -24,21 +24,21 @@ public static class Comparer
         public static EqualityComparer<T, TResult> ContraMap<T, TResult>(this IEqualityComparer<T> comparer, Func<TResult, T> contraMapper)
```
Comment (atifaziz) [inline src/Core/Comparer.cs]: You can return `IEqualityComparer<TResult>` simply.

File: src/Core/Comparer.cs [hunk 7]
```diff
@@ -24,21 +24,21 @@ public static class Comparer
...
             => new EqualityComparer<T, TResult>(comparer, contraMapper);
 
-        sealed public class EqualityComparer<T, TResult> : IEqualityComparer<TResult>
+        public sealed class EqualityComparer<T, TResult> : IEqualityComparer<TResult>
```
Comment (atifaziz) [inline src/Core/Comparer.cs]: `EqualityComparer<T, TResult>` doesn't need to be public.

File: src/Core/Comparer.cs [hunk 8]
```diff
@@ -24,21 +24,21 @@ public static class Comparer
...
         {
-            Func<TResult, T> contraMapper;
-            IEqualityComparer<T> comparer;
+            readonly Func<TResult, T> _contraMapper;
+            readonly IEqualityComparer<T> _comparer;
 
             public EqualityComparer(IEqualityComparer<T> comparer, Func<TResult, T> contraMapper)
             {
                 if (contraMapper == null) throw new ArgumentNullException(nameof(contraMapper));
```
Comment (atifaziz) [inline src/Core/Comparer.cs]: The should be in the order of the parameters.

File: src/Core/Comparer.cs [hunk 9]
```diff
@@ -24,21 +24,21 @@ public static class Comparer
         public static EqualityComparer<T, TResult> ContraMap<T, TResult>(this IEqualityComparer<T> comparer, Func<TResult, T> contraMapper)
             => new EqualityComparer<T, TResult>(comparer, contraMapper);
```
Comment (atifaziz) [inline src/Core/Comparer.cs]: Move arrow to the end of the function declaration to be consistent with the rest of the project.

File: src/Core/Comparer.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+//
+#endregion
+
+namespace WebLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    public static class Comparer
```
Comment (atifaziz) [inline src/Core/Comparer.cs]: Rename `Comparer` to `EqualityComparerExtensions`.

Comment (gmpl) [inline src/Core/Comparer.cs]: I renamed it to <code>ComparerExtensions</code> because we can include the same method for <code>IComparer</code> there.

**Comment by atifaziz** [comment]
Merged into `master` by d83265a3eaf9db68b556230a0475ef23e46a51e4

**Comment by atifaziz** [comment]
I realised it a bit too late but unfortunately this is conceptually flawed as it stands. Which state gets delivered as part of the output? Any kind of aggregation would cause a problem.

**Pull request closed without merging**